gusto/physics.py

```diff
@@ -84,62 +97,285 @@ def __init__(self, state):
         dot_r_cond = ((self.water_v - w_sat) /
                       (dt * (1.0 + ((L_v ** 2.0 * w_sat) /
                                     (cp * R_v * T ** 2.0)))))
-        dot_r_cond = self.water_v - w_sat
 
         # make cond_rate function, that needs to be the same for all updates in one time step
-        self.cond_rate = Function(Vt)
-        self.cond_rate = state.fields('cond_rate', Vt)
+        cond_rate = Function(Vt)
 
         # adjust cond rate so negative concentrations don't occur
         self.lim_cond_rate = Interpolator(conditional(dot_r_cond < 0,
                                                       max_value(dot_r_cond, - self.water_c / dt),
-                                                      min_value(dot_r_cond, self.water_v / dt)), self.cond_rate)
+                                                      min_value(dot_r_cond, self.water_v / dt)), cond_rate)
 
         # tell the prognostic fields what to update to
-        self.water_v_new = Interpolator(self.water_v - dt * self.cond_rate, Vt)
-        self.water_c_new = Interpolator(self.water_c + dt * self.cond_rate, Vt)
+        self.water_v_new = Interpolator(self.water_v - dt * cond_rate, Vt)
+        self.water_c_new = Interpolator(self.water_c + dt * cond_rate, Vt)
         self.theta_new = Interpolator(self.theta *
-                                      (1.0 + dt * self.cond_rate *
+                                      (1.0 + dt * cond_rate *
                                        (cv * L_v / (c_vml * cp * T) -
                                         R_v * cv * c_pml / (R_m * cp * c_vml))), Vt)
 
     def apply(self):
         self.rho_broken.assign(self.rho_interpolator.interpolate())
         self.rho_recoverer.project()
-        self.lim_cond_rate.interpolate()
-        self.theta.assign(self.theta_new.interpolate())
-        self.water_v.assign(self.water_v_new.interpolate())
-        self.water_c.assign(self.water_c_new.interpolate())
+        for i in range(self.iterations):
+            self.lim_cond_rate.interpolate()
+            self.theta.assign(self.theta_new.interpolate())
+            self.water_v.assign(self.water_v_new.interpolate())
+            self.water_c.assign(self.water_c_new.interpolate())
 
 
 class Fallout(Physics):
     """
     The fallout process of hydrometeors.
 
     :arg state :class: `.State.` object.
+    :arg moments: an integer denoting which rainfall scheme to use.
```

Comment:
I was trying to think if there are more intuitive values for this argument but it's so complex that I'm not sure there is an obvious answer

Reply:
Use an Enum. 

AdvectedMoments.MEAN

And similar

Reply:
https://docs.python.org/3/library/enum.html

Reply:
I don't think I understand enum well enough to see how to use it here!

Reply:
It just removes the need for magic numbers in the code.

```python
class AdvectedMoments(Enum):
   M0 = 0
   M3 = 1
   # any other moments here
```

```python
Fallout(..., moments=AdvectedMoments.M0, ...)
```

Now the value describes what is going on.

Reply:
I've tried to make this change now, possibly I still haven't exactly understood how to do it properly